### ISSUE TITLE
64 bit on osx bin files won't compile.

### ISSUE THREAD

**Issue by Sophrinix** [liuliu/ccv#12]
I am using OSX lion( 10.7.4) on a i7 quad core. 
When I go to the lib directory and ./configure  and make (choosing clang)
I am able to successfully compile from that make file.
However when I move into the bin directory to compile the c files in that directory this is what follows:

  ➜  bin git:(unstable) make
make -C ../lib

 USE: clang
 COMPILE FLAGS: -D HAVE_LIBJPEG
 LINK FLAGS: -ljpeg

make[2]: Nothing to be done for `all'.
clang bbffmt.c -o bbffmt.o -c -O3 -Wall -I../lib -D HAVE_LIBJPEG
clang -o bbffmt bbffmt.o -L../lib -lccv -ljpeg -lm
Undefined symbols for architecture x86_64:
  "_isnanf", referenced from:
      _ccv_any_nan in libccv.a(ccv_util.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: **\* [bbffmt] Error 1
➜  bin git:(unstable) 

I feel like this is just a simple flag issue in the makefile. how do i get it to recognize 64bit ness or will i have to pass a i386 flag?

**Comment by Sophrinix** [comment]
one more note:
when I do a make clean (from bin) and a make all this is what happens:

➜  bin git:(unstable) make all
make -C ../lib

 USE: clang
 COMPILE FLAGS: -D HAVE_LIBJPEG
 LINK FLAGS: -ljpeg

clang ccv_cache.c -o ccv_cache.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang ccv_memory.c -o ccv_memory.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang 3rdparty/sha1/sha1.c -o 3rdparty/sha1/sha1.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang 3rdparty/kissfft/kiss_fft.c -o 3rdparty/kissfft/kiss_fft.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang 3rdparty/kissfft/kiss_fftnd.c -o 3rdparty/kissfft/kiss_fftnd.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang 3rdparty/kissfft/kiss_fftr.c -o 3rdparty/kissfft/kiss_fftr.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang 3rdparty/kissfft/kiss_fftndr.c -o 3rdparty/kissfft/kiss_fftndr.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang 3rdparty/kissfft/kissf_fft.c -o 3rdparty/kissfft/kissf_fft.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang 3rdparty/kissfft/kissf_fftnd.c -o 3rdparty/kissfft/kissf_fftnd.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang 3rdparty/kissfft/kissf_fftr.c -o 3rdparty/kissfft/kissf_fftr.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang 3rdparty/kissfft/kissf_fftndr.c -o 3rdparty/kissfft/kissf_fftndr.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang ccv_io.c -o ccv_io.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang ccv_numeric.c -o ccv_numeric.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang ccv_algebra.c -o ccv_algebra.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang ccv_util.c -o ccv_util.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
ccv_util.c:77: 8 : warning: implicit declaration of function 'isnanf' is invalid in C99 [-Wimplicit-function-declaration]
                        if (isnanf(da->data.f32[i]))
                            ^
1 warning generated.
clang ccv_basic.c -o ccv_basic.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG

clang ccv_classic.c -o ccv_classic.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang ccv_daisy.c -o ccv_daisy.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang ccv_sift.c -o ccv_sift.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang ccv_bbf.c -o ccv_bbf.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
ccv_bbf.c:371:28 : warning: unused function '_ccv_bbf_genetic_qsort' [-Wunused-function]
static CCV_IMPLEMENT_QSORT(_ccv_bbf_genetic_qsort, ccv_bbf_gene_t, less_than)
                           ^
./ccv_internal.h:526:28: note: expanded from macro 'CCV_IMPLEMENT_QSORT'
    CCV_IMPLEMENT_QSORT_EX(func_name, T, cmp, _ccv_qsort_default_swap, int)
                           ^
./ccv_internal.h:362: 6: note: expanded from macro 'CCV_IMPLEMENT_QSORT_EX'
void func_name(T _array, size_t total, user_data_type aux)                                      \
     ^
ccv_bbf.c:550:23: warning: unused function '_ccv_bbf_best_gene' [-Wunused-function]
static ccv_bbf_gene_t _ccv_bbf_best_gene(ccv_bbf_gene_t_ gene, int pnum, int point_min, unsigned char*\* posdata, int posnum, ...
                      ^
2 warnings generated.
clang ccv_swt.c -o ccv_swt.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
clang ccv_dpm.c -o ccv_dpm.o -c -O3 -ffast-math -mtune=native -Wall -D HAVE_LIBJPEG
ar rcs libccv.a ccv_cache.o ccv_memory.o 3rdparty/sha1/sha1.o 3rdparty/kissfft/kiss_fft.o 3rdparty/kissfft/kiss_fftnd.o 3rdparty/kissfft/kiss_fftr.o 3rdparty/kissfft/kiss_fftndr.o 3rdparty/kissfft/kissf_fft.o 3rdparty/kissfft/kissf_fftnd.o 3rdparty/kissfft/kissf_fftr.o 3rdparty/kissfft/kissf_fftndr.o ccv_io.o ccv_numeric.o ccv_algebra.o ccv_util.o ccv_basic.o ccv_classic.o ccv_daisy.o ccv_sift.o ccv_bbf.o ccv_swt.o ccv_dpm.o
clang bbffmt.c -o bbffmt.o -c -O3 -Wall -I../lib -D HAVE_LIBJPEG
clang -o bbffmt bbffmt.o -L../lib -lccv -ljpeg -lm
Undefined symbols for architecture x86_64:
  "_isnanf", referenced from:
      _ccv_any_nan in libccv.a(ccv_util.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: **\* [bbffmt] Error 1

**EDIT: formatting**

**Comment by Sophrinix** [comment]
turns out that this isn't a problem if you dont use -sse3. However 64 bit ness is still an issue:

clang -o bbffmt bbffmt.o -L../lib -lccv -ljpeg -lpng -lz -lgsl -lm
Undefined symbols for architecture x86_64:
  "_isnanf", referenced from:
      _ccv_any_nan in libccv.a(ccv_util.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

however osx lion 10.7 does have a problem with libpng. 
It cannot find these constants,  Z_HUFFMAN_ONLY, MAX_MEM_LEVEL, Z_BEST_SPEED

The answer is suppose to call ENV.x11

TL;DR: get your act together apple.

**Comment by liuliu** [comment]
originally, I put -mtune=native in the compile script, it seems to be a problem. just pushed a new version that should fixed various compiling errors/warnings with regards to Mac OSX.

**Comment by Sophrinix** [comment]
this completely fixed it. 

the only thing left to make this a smooth install would be to get all the dependencies in an easy to install format.
I'm thinking about a homebrew package. However, for right now it is ultimately more useful under linux.

thank you very much for your work on this project.

**Comment by maelp** [comment]
I still have this problem on os x Lion
(on both unstable and stable branches)

**Comment by maelp** [comment]
Specifically, I have the following errors during library build

```
ccv_io.c:144:8: warning: implicit declaration of function 'fmemopen' is invalid in C99 [-Wimplicit-function-declaration]
                fd = fmemopen((void*)in, (size_t)rows, "rb");
                     ^
ccv_io.c:144:6: warning: incompatible integer to pointer conversion assigning to 'FILE *' (aka 'struct __sFILE *') from 'int'; 
                fd = fmemopen((void*)in, (size_t)rows, "rb");
                   ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ccv_dpm.c:444:9: warning: incompatible pointer types assigning to 'double *' from 'int *' [-Wincompatible-pointer-types]
        prob.y = (int*)malloc(sizeof(int) * (cnum + negex->rnum) * (!!symmetric + 1));
               ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

and this is what I obtain when trying to make the binaries:

```
Undefined symbols for architecture x86_64:
  "_fmemopen", referenced from:
      _ccv_read_impl in libccv.a(ccv_io.o)
ld: symbol(s) not found for architecture x86_64
```

**Comment by liuliu** [comment]
@maelp this is another problem, in 0.2, I use fmemopen to open in-memory block as file descriptor to read them in. This is not supported on Mac OSX, I pushed a fix for that to stable. Thanks.

**Comment by liuliu** [comment]
also, the new liblinear version has y defined as double*, will fix that too.
